File: services/fifo.go
package services

type FIFOLot struct {
	Side     int32   // 1 = buy, -1 = sell
	Quantity float32
	Price    float32
}

type Position struct {
	Quantity    float32 // remaining units
	AvgCost     float32 // per-unit cost of remaining units
	CostBasis   float32 // total cost of remaining units
	RealizedPnl float32
}

// ComputeFIFO processes buy/sell lots FIFO. A sell consumes the oldest
// remaining buy lots. Realized P&L = sell proceeds - FIFO cost of sold units.
// Over-sells leak a negative quantity (the caller rejects them).
func ComputeFIFO(lots []FIFOLot) Position {
	var pos Position
	type buyLot struct {
		quantity float32
		price    float32
	}
	var buys []buyLot

	for _, lot := range lots {
		if lot.Side >= 0 {
			buys = append(buys, buyLot{quantity: lot.Quantity, price: lot.Price})
			continue
		}
		remaining := lot.Quantity
		for remaining > 0 && len(buys) > 0 {
			if buys[0].quantity <= remaining {
				pos.RealizedPnl += (lot.Price - buys[0].price) * buys[0].quantity
				remaining -= buys[0].quantity
				buys = buys[1:]
			} else {
				pos.RealizedPnl += (lot.Price - buys[0].price) * remaining
				buys[0].quantity -= remaining
				remaining = 0
			}
		}
		if remaining > 0 {
			pos.Quantity -= remaining // over-sell leaks negative
		}
	}

	for _, b := range buys {
		pos.Quantity += b.quantity
		pos.CostBasis += b.quantity * b.price
	}
	if pos.Quantity > 0 {
		pos.AvgCost = pos.CostBasis / pos.Quantity
	}
	return pos
}

File: src/service/fifo.rs
//! FIFO cost-basis math, mirroring Go's `services/fifo.go`.

#[derive(Debug, Clone, Copy)]
pub struct FifoLot {
    pub side: i64, // 1 buy, -1 sell
    pub quantity: f64,
    pub price: f64,
}

#[derive(Debug, Clone, Copy, Default)]
pub struct Position {
    pub quantity: f64,
    pub avg_cost: f64,
    pub cost_basis: f64,
    pub realized_pnl: f64,
}

/// Process buy/sell lots FIFO. A sell consumes the oldest remaining buy lots.
/// Realized P&L = sell proceeds - FIFO cost of sold units. Over-sells leak a
/// negative quantity (the caller rejects them).
pub fn compute_fifo(lots: &[FifoLot]) -> Position {
    let mut pos = Position::default();
    let mut buys: Vec<(f64, f64)> = Vec::new(); // (quantity, price)

    for lot in lots {
        if lot.side >= 0 {
            buys.push((lot.quantity, lot.price));
            continue;
        }
        let mut remaining = lot.quantity;
        while remaining > 0.0 && !buys.is_empty() {
            let (b_qty, b_price) = buys[0];
            if b_qty <= remaining {
                pos.realized_pnl += (lot.price - b_price) * b_qty;
                remaining -= b_qty;
                buys.remove(0);
            } else {
                pos.realized_pnl += (lot.price - b_price) * remaining;
                buys[0].0 -= remaining;
                remaining = 0.0;
            }
        }
        if remaining > 0.0 {
            pos.quantity -= remaining; // over-sell leaks negative
        }
    }

    for (b_qty, b_price) in buys {
        pos.quantity += b_qty;
        pos.cost_basis += b_qty * b_price;
    }
    if pos.quantity > 0.0 {
        pos.avg_cost = pos.cost_basis / pos.quantity;
    }
    pos
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn fifo_buy_sell_consumes_oldest() {
        let pos = compute_fifo(&[
            FifoLot { side: 1, quantity: 10.0, price: 10.0 },
            FifoLot { side: 1, quantity: 10.0, price: 20.0 },
            FifoLot { side: -1, quantity: 15.0, price: 25.0 },
        ]);
        assert_eq!(pos.quantity, 5.0);
        assert_eq!(pos.avg_cost, 20.0); // remaining 5 from the 20-cost lot
        assert_eq!(pos.realized_pnl, (25.0 - 10.0) * 10.0 + (25.0 - 20.0) * 5.0);
    }

    #[test]
    fn fifo_partial_sell() {
        let pos = compute_fifo(&[
            FifoLot { side: 1, quantity: 10.0, price: 10.0 },
            FifoLot { side: -1, quantity: 4.0, price: 15.0 },
        ]);
        assert_eq!(pos.quantity, 6.0);
        assert_eq!(pos.realized_pnl, (15.0 - 10.0) * 4.0);
    }

    #[test]
    fn fifo_empty() {
        let pos = compute_fifo(&[]);
        assert_eq!(pos.quantity, 0.0);
        assert_eq!(pos.avg_cost, 0.0);
    }
}
